django/db/models/fields/__init__.py

```diff
@@ -2395,3 +2291,113 @@ def formfield(self, **kwargs):
             'form_class': forms.UUIDField,
             **kwargs,
         })
+
+
+class AutoFieldMixin:
+
+    def __init__(self, *args, **kwargs):
+        kwargs['blank'] = True
+        super().__init__(*args, **kwargs)
+
+    def check(self, **kwargs):
+        return [
+            *super().check(**kwargs),
+            *self._check_primary_key(),
+        ]
+
+    def _check_primary_key(self):
+        if not self.primary_key:
+            return [
+                checks.Error(
+                    'AutoFields must set primary_key=True.',
+                    obj=self,
+                    id='fields.E100',
+                ),
+            ]
+        else:
+            return []
+
+    def deconstruct(self):
+        name, path, args, kwargs = super().deconstruct()
+        del kwargs['blank']
+        kwargs['primary_key'] = True
+        return name, path, args, kwargs
+
+    def validate(self, value, model_instance):
+        pass
+
+    def get_db_prep_value(self, value, connection, prepared=False):
+        if not prepared:
+            value = self.get_prep_value(value)
+            value = connection.ops.validate_autopk_value(value)
+        return value
+
+    def get_prep_value(self, value):
+        from django.db.models.expressions import OuterRef
+        return value if isinstance(value, OuterRef) else super().get_prep_value(value)
+
+    def contribute_to_class(self, cls, name, **kwargs):
+        assert not cls._meta.auto_field, (
+            "Model %s can't have more than one auto-incrementing field."
```

Comment:
So `auto-incrementing` implies that this is numeric. Given we're aiming toward non-integer support, do we want to rephrase this slightly to avoid needing to change it again later? I'm conscious that support for generated columns is also on the horizon, so we need to be careful with using the word "generated" here too, but "auto-generated" _may_ be OK...

I'll leave this to you to change if you see fit prior to merge.

Reply:
I don't think that we won't have this restriction anymore for non-integer auto-field (at least not in all databases because PostgreSQL and Oracle support `RETURNING` clause with multiple fields). Let's leave `auto-generated` for now.